docs/src/developing/runtime-facilities/zk-token-proof.md

```diff
@@ -0,0 +1,123 @@
+
+### Pedersen commitments and ElGamal encryption
+
+The ZK Token proof program verifies zero-knowledge proofs for Pedersen
+commitments and ElGamal encryption, which are common cryptographic primitives
+that are incorporated in many existing cryptographic protocols.
+
+ElGamal encryption is a popular instantiation of a public-key encryption scheme.
+An ElGamal keypair consists of an ElGamal public key and an ElGamal secret key.
+Messages can be encrypted under a public key to produce a ciphertext. A
+ciphertext can then be decrypted using a corresponding ElGamal secret key. The
+variant that is used in the proof program is the
+[twisted ElGamal encryption](https://eprint.iacr.org/2019/319) over the elliptic
+curve [curve25519](https://www.rfc-editor.org/rfc/rfc7748#section-4.1).
+
+The Pedersen commitment scheme is a popular instantiation of a cryptographic
+commitment scheme. A commitment scheme allows a user to wrap a message into a
+commitment with a purpose of revealing the committed message later on. Like a
+ciphertext, the resulting commitment does not reveal any information about the
+containing message. At the same time, the commitment is binding in that the user
+cannot change the original value that is contained in a commitment.
+
+Interested readers can refer to the following resources for a more in-depth
+treatment of Pedersen commitment and the (twisted) ElGamal encryption schemes.
+
+- [Notes](./zk-docs/twisted_elgamal.pdf) on the twisted ElGamal encryption
+- A technical
+  [overview](https://github.com/solana-labs/solana-program-library/blob/master/token/zk-token-protocol-paper/part1.pdf)
+  of the SPL Token 2022 confidential extension
+- Pretty Good Confidentiality [research paper](https://eprint.iacr.org/2019/319)
+
+The ZK Token proof program contains proof verification instructions on various
+zero-knowledge proofs for working with the Pedersen commitment and ElGamal
+encryption schemes. For example, the `VerifyRangeProofU64` instruction verifies
+a zero-knowledge proof certifying that a Pedersen commitment contains an
+unsigned 64-bit number as the message. The `VerifyPubkeyValidity` instruction
+verifies a zero-knowledge proof certifying that an ElGamal public key is a
+properly formed public key.
+
+### Context Data
+
+The proof data associated with each of the ZK Token proof instructions are
+logically divided into two parts:
+
+- The <em>context</em> component contains the data that a zero-knowledge proof
+  is certifying. For example, context component for a `VerifyRangeProofU64`
+  instruction data is the Pedersen commitment that holds an unsigned 64-bit
+  number. The context component for a `VerifyPubkeyValidity` instruction data is
+  the ElGamal public key that is properly formed.
+- The <em>proof</em> component contains the actual mathematical pieces that
+  certifies different properties of the context data.
```

Comment:
```suggestion
  certify different properties of the context data.
```

Reply:
Ah sorry about that... I should have double checked and caught all these mistakes before pinging :pray: ...